crates/services/sync/src/state.rs

```diff
@@ -0,0 +1,212 @@
+pub struct State {
+    status: Status,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+/// Status of the sync service.
+pub enum Status {
+    /// The service is not initialized and there is nothing to sync.
+    Uninitialized,
+    /// This range is being processed.
+    Processing(RangeInclusive<u32>),
+    /// This height is committed.
+    Committed(u32),
+}
+
+impl State {
+    /// Create a new state from the current committed and observed heights.
+    pub fn new(
+        committed: impl Into<Option<u32>>,
+        observed: impl Into<Option<u32>>,
+    ) -> Self {
+        let status = match (committed.into(), observed.into()) {
+            // Both the committed and observed heights are known.
+            (Some(committed), Some(observed)) => {
+                // If there is a gap between the committed and observed heights,
+                // the service is processing the gap otherwise the service is
+                // has nothing to sync.
+                committed
+                    .checked_add(1)
+                    .map_or(Status::Committed(committed), |next| {
+                        let range = next..=observed;
+                        if range.is_empty() {
+                            Status::Committed(committed)
+                        } else {
+                            Status::Processing(range)
+                        }
+                    })
+            }
+            // Only the committed height is known, so the service has nothing to sync.
+            (Some(committed), None) => Status::Committed(committed),
+            // Only the observed height is known, so the service is processing
+            // up to that height.
+            (None, Some(observed)) => Status::Processing(0..=observed),
+            // No heights are known, so the service is uninitialized.
+            (None, None) => Status::Uninitialized,
+        };
+        Self { status }
+    }
+
+    /// Get the current range to process.
+    pub fn process_range(&self) -> Option<RangeInclusive<u32>> {
+        match &self.status {
+            Status::Processing(range) => Some(range.clone()),
+            _ => None,
+        }
+    }
+
+    /// Record that a block has been committed.
+    pub fn commit(&mut self, height: u32) {
+        let new_status = match &self.status {
+            // Currently processing a range and recording a commit.
+            Status::Processing(range) => match height.cmp(range.end()) {
+                // The commit is less than the end of the range, so the range
+                // is still being processed.
+                std::cmp::Ordering::Less => {
+                    Some(Status::Processing(height.saturating_add(1)..=*range.end()))
+                }
+                // The commit is equal or greater than the end of the range,
+                // so the range is fully committed.
+                std::cmp::Ordering::Equal | std::cmp::Ordering::Greater => {
```

Comment:
Importing `std::cmp::Ordering` would make this cleaner since it is used multiple times